### PR TITLE
Display better error messages for E0282

### LINKED ISSUE #36554: error 282: If we have no info about the actual type param, say something about the formal type param
e.g.,
```
unable to infer enough type information about `_`
```

and inline: `cannot infer type for \`_\``

This is when calling a function with a generic type parameter. The actual type parameter is inferred, so we can't name it, but there is a corresponding formal parameter we could name.

E.g., if calling `foo<T>(...)` we have `foo(...)` and so get a type error for `_`, but we could mention `T` here, something like `cannot infer type for \`_\`, the inferred type corresponding to `T` in this call to `foo``.

### PR DESCRIPTION
**Pull request by KiChjang**
Fixes #36554.

### CODE REVIEW

**Comment by frewsxcv** [comment]
FYI, Travis found some issues.

File: src/librustc/infer/lattice.rs [hunk 1]
```diff
@@ -64,14 +65,14 @@ pub fn super_lattice_tys<'a, 'gcx, 'tcx, L>(this: &mut L,
     match (&a.sty, &b.sty) {
         (&ty::TyInfer(TyVar(..)), &ty::TyInfer(TyVar(..)))
             if infcx.type_var_diverges(a) && infcx.type_var_diverges(b) => {
-            let v = infcx.next_diverging_ty_var();
+            let v = infcx.next_diverging_ty_var(TypeVariableOrigin::LatticeVariable);
```
Comment (nikomatsakis) [inline src/librustc/infer/lattice.rs]: We can give more information about lattice variables -- at minimum, we can give a span. The `this` variable is always of a type (either `Lub` or `Glb`) that has access to a `CombineFields` (`self.fields`, in each case). `CombineFields` has a field `trace` of type `TypeTrace`, and hiding in there is a field `cause` of type `ObligationCause`. This is basically "why we are computing the common-supertype or common-subtype of these types". At minimum though it'd be nice to add a method `fn cause(&self) -> &OlbligationCause<'tcx>` to `LatticeDir` that returns `&self.fields.cause`. Then we can create give a span here like `LatticeVariable(this.cause().span)`.

File: src/librustc/infer/type_variable.rs [hunk 2]
```diff
@@ -28,8 +29,24 @@ pub struct TypeVariableTable<'tcx> {
...
+    NormalizeProjectionType(Span),
+    TypeInference(Span),
+    TypeParameterDefinition(Span, ast::Name),
+    TransformedUpvar(Span),
+    SubstitutionPlaceholder(Span),
+    AutoDeref(Span),
+    AdjustmentType(Span),
+    DivergingStmt(Span),
+    DivergingBlockExpr(Span),
+    LatticeVariable,
```
Comment (nikomatsakis) [inline src/librustc/infer/type_variable.rs]: ...if we give `LatticeVariable` a span, then every variant will have one.

**Review by nikomatsakis — CHANGES_REQUESTED** [review]
Looks good! I just left one minor suggestion. The other thing that I'm missing is some kind of dedicated test where we show off a nice error. Maybe a UI test for this case?

```rust
fn foo<X>() { }
fn bar() {
    foo();
}
```

**Comment by KiChjang** [comment]
@nikomatsakis There is actually a UI test that demonstrates this and it's at `src/test/ui/codemap_tests/repair_span_std_macros.rs`. Would that suffice or do you think it's still better to create a new one specifically for this?

File: src/librustc/infer/lattice.rs (hunk 1 above)
Comment (KiChjang) [inline src/librustc/infer/lattice.rs]: Ack, the complication here is that `self.fields.cause` is an `Option<ObligationCause>`. How should I ensure that there must always be an `ObligationCause`?

Comment (KiChjang) [inline src/librustc/infer/lattice.rs]: Found it, `self.trace.cause` may work.

**Comment by nikomatsakis** [comment]
@KiChjang 

> There is actually a UI test that demonstrates this and it's at src/test/ui/codemap_tests/repair_span_std_macros.rs. Would that suffice or do you think it's still better to create a new one specifically for this?

I think it's better to have a targeted test. For one thing, when doing refactorings, it is often hard to tell if a change to the test's output is acceptable or not; but if there is a focused test that is aimed at this particular feature, it will be more obvious. Another thing that happens from time to time is that when we are doing refactorings some test is hard to port to the new system and winds up getting deleted, thereby inadvertendly removing the only test for *this* feature. So I think it's best to have simple tests that specifically target your change (in addition to the ones which accidentally do so).

Plus, the scenario in that UI test was not quite the one I suggested, right?

```
fn foo<X>() { }
fn bar() {
    foo();
}
```

**Comment by nikomatsakis** [comment]
(Sorry for the delay.)

**Comment by KiChjang** [comment]
UI test added.

**Comment by nikomatsakis** [comment]
@KiChjang 

tidy error: `/checkout/src/test/ui/missing-items/missing-type-parameter.rs: incorrect license`

**Review by nikomatsakis — APPROVED** [review]

**Comment by KiChjang** [comment]
Dumb vim. Fixed license now.

**Comment by KiChjang** [comment]
Huh... I'm not sure what the error in Travis CI is about this time.

**Pull request merged**